### DISCUSSION TITLE
Handy Hint: LUA external files

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#177] [Show and tell]
Within the ctrlr LUA editor, there is a command under the file menu "convert to files"

My recommendations to all programmers is to use this.  It saves all of your LUA methods as external files, and if you have your project synced to a repository on github or similar, it allows for version control and all of the normal github functionality.

It also means that your files are kept outside of Ctrlr, so if you do manage to crash Ctrlr, or otherwise stuff up your project so it can't be restarted, you can edit the .lua file offline to fix it. Also, if it ctrlr refuses to open your project (it has happened to me on more than one occasion), you can still access all of your work and rebuild things quickly.

You will need to repeat this each time you add a new method, and what(), how() etc seem to keep getting new versions added, which you can delete, but I'm finding it a much nicer way to develop, in conjunction with Visual  Studio Code, than just using LUA.  I use Visual Studio to handle version management, but code in the Ctrlr LUA editor. Coding in Visual Studio means that you need to reopen your panel to refresh the code, and you need to compile within Ctrlr anyway, so it makes sense to do it this way.

**Comment by dobo365** [comment]
Thx a lot for this tip that I didn't know.
Will definitely try it!

**Comment by bijlevel** [comment]
Hmm, looked everywhere but can't find it in my version of Ctrlr (5.3.122 on Macintosh), only 'export as xml' among the export options.

**Reply by dobo365** [reply]
One thing: the last stable version on Mac is 5.3.198.
Second thing: I'm in 5.3.201 (last stable on PC) and indeed don't find that option "Convert to files"

**Reply by bijlevel** [reply]
I know, but with versions above 5.3.122 I had problems with displaying resources added in lower versions. But maybe that was fixed in 5.3.198?

**Comment by dobo365** [comment]
@Godlike-Productions Which Ctrlr version are you using? 
This seems not available in the last stable ones (5.3.201 / 5.3.198).
Thx!

**Comment by Godlike-Productions** [comment]
I'm in 6.02.26 (which is actually before the latest builds I think. They reverted back to V5.

The command is in the LUA editor
![coverttofiles](https://user-images.githubusercontent.com/13851825/108621264-a9439280-746c-11eb-8ef5-aa4993bd05ed.PNG)

**Reply by bijlevel** [reply]
It is indeed present on the Windows version 6.0.26, but a quick test of this Win version with a panel made on a Mac did fail, so I'm sticking with the Mac version until my panel is finished ;-)

<img width="618" alt="Screenshot 2021-02-21 at 11 25 44" src="https://user-images.githubusercontent.com/11859363/108622287-083ae480-7438-11eb-877e-4f326e4baf16.png">

**Comment by shooking** [comment]
I used this a lot in the past - but still things can get messed up. It's like one morning Ctrlr wakes up with the "right arse" and decides to hide stuff. I also simply put my panel into git and git commit it periodically.

Working outside of Ctrl editor is a lot easier on a Pi simply because the fonts are so large. I have not managed to fix that yet - reverse engineering Zoom pedals takes my time.

Version = 9597fad0.0, Build date = Wed 06 Jan 2021 03:13:42 PM GMT, Branch = Stable, Juce = 5.4.1, libusb = 1.0.19, liblo = 0.28, lua = Lua 5.1.4, luabind = 0.9.0, boost = 1.67.0,

I do hope Roman starts exposing the release tags so I can know what I am actually building.

CAREFUL with 6.0 - the MemoryBlocks changed semantics. Caught me out big time. Thank heavens for Dnaldoog is all I can say. He helped explain the semantics.

**Reply by Godlike-Productions** [reply]
Out of curiosity what changed with MemoryBlocks?  I haven't come accross any issues, and I'm using them pretty extensively.  Just wondering what to look out for.

**Reply by shooking** [reply]
https://ctrlr.org/forums/topic/memoryblock-v-ctrlrluamemoryblock/
